Create plugin for yamlTest task

jakelandis opened the pull request:
This commit creates a new Gradle plugin to provide a separate task name
and source set for running YAML based REST tests. The only project
converted to use the new plugin in this PR is distribution/archives/integ-test-zip.
For which the testing has been moved to :rest-api-spec since it makes the most 
sense and it avoids a small but awkward change to the distribution plugin. 

The remaining cases in modules, plugins, and x-pack will be handled in followups.

This plugin is distinctly different from the plugin introduced in #55896 since
the YAML REST tests are intended to be black box tests over HTTP. As such they
should not (by default) have access to the classpath for that which they are testing.

The YAML based REST tests will be moved to separate source sets (yamlRestTest). 
The which source is the target for the test resources is dependent on if this
new plugin is applied. If it is not applied, it will default to the test source
set. 

Further, this introduces a breaking change for plugin developers that
use the YAML testing framework. They will now need to either use the new source set 
and matching task, or configure the rest resources to use the old "test" source set that 
matches the  old integTest task. (The former should be preferred). 

As part of this change (which is also breaking for plugin developers) the  
rest resources plugin has been removed from the build plugin and now requires
either explicit application or application via the new YAML REST test plugin. 

Plugin developers should be able to fix the breaking changes to the YAML tests 
by adding `apply plugin: 'elasticsearch.yaml-rest-test'` and moving the YAML tests
under a `yamlRestTest` folder (instead of `test`)

jakelandis committed:
introduce plugin

jakelandis committed:
clean up

jakelandis committed:
add another todo

jakelandis committed:
clean up comment

jakelandis committed:
fix runner setup

jakelandis commented:
@elasticmachine test this please

mark-vieira commented:
> it is still desirable for plugin development to support
YAML based REST tests from the standard Java based directories.

Why? The main motivator here is to make these things very explicit. The directory this stuff goes in dictates the kind of test. Supporting such a "hybrid" approach conflicts with that goal. In what scenario would a developer require these test live in the unit testing source set instead of the YAML test one?

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/RestIntegTestTask.java:
```diff
@@ -36,6 +36,8 @@
     private static final String TESTS_CLUSTER_NAME = "tests.clustername";
 
+    // TODO: refactor this so that work is not done in constructor and find usages and register them, not create them
+    // See: https://docs.gradle.org/current/userguide/task_configuration_avoidance.html
```
FWIW, there's an open issue for this. Perhaps we can link it here.

https://github.com/elastic/elasticsearch/issues/47804

jakelandis replied:
done.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/CopyRestApiTask.java:
```diff
@@ -109,7 +111,8 @@ public FileTree getInputDir() {
     @OutputDirectory
     public File getOutputDir() {
-        return new File(getTestSourceSet().getOutput().getResourcesDir(), REST_API_PREFIX);
+        assert Util.getTestSourceSet(getProject()).isPresent();
```
We should use `orElseThrow()` here and supply an exception with an appropriate message.

jakelandis replied:
I added a message to the `assert`

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/util/Util.java:
```diff
@@ -129,15 +129,39 @@ public static FileTree getJavaTestAndMainSourceResources(Project project, Action
      * @param project The project to look for test Java resources.
      * @return An Optional that contains the Java test SourceSet if it exists.
      */
+    public static Optional<SourceSet> getTestSourceSet(Project project) {
```
As mentioned in my comment, I don't think we need to support this case. We should _require_ that all YAML tests live in the appropriate source set. This awkward conditional stuff is a great example of why.

Also, what happens if a project has both tests types? Which source set should this method return? As part of splitting these types of tests up there is no longer a canonical "test source set", as any given project might (and likely will) have several.

jakelandis replied:
Discussed in person. Since this is used primarily for copy around of REST resources, I will remove this method in favor of having the REST resources be explicit about which source set they will use dependent on the application (or not) of the new YAML test task. 

Also, this PR updated to use this with test conventions task, but that isn't needed for custom test source sets.

There will still be a todo to circle back around and remove the conditional aspect of which test source set is used for copying around REST resources, but that can only be done after all existing usages of the java test source set have been addressed. This includes ensuring that the clients teams won't break on moving tests around, as well all internal usages where YAML tests can live in the java test source set. While plugin developers may be non-passively impacted by requiring the new YAML plugin, the migration path is pretty simple and would only impact those that leverage YAML testing (suspected to be very few).

jakelandis replied:
done ... will add summary to main comment.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlTestPlugin.java:
```diff
@@ -0,0 +1,126 @@
+    @Override
+    public void apply(Project project) {
+
+        if (project.getPluginManager().hasPlugin("elasticsearch.build") == false
```
We shouldn't have to do this. Let's just apply the correct plugin here, which is going to be the `ElasticsearchJavaPlugin` most likely. We don't want to have users have to worry about this.

jakelandis replied:
done

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlTestPlugin.java:
```diff
@@ -0,0 +1,126 @@
+        yamlTestTask.setDescription("Runs the YAML based REST tests against an external cluster");
+
+        // setup task dependency
+        project.getDependencies().add(SOURCE_SET_NAME + "Compile", project.project(":test:framework"));
```
For now we need to check if `BuildParams.isInternal` here and use an artifact dependency if not.

jakelandis replied:
done

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlTestPlugin.java:
```diff
@@ -0,0 +1,126 @@
+        runner.setTestClassesDirs(yamlTestSourceSet.getOutput().getClassesDirs());
+        runner.setClasspath(yamlTestSourceSet.getRuntimeClasspath());
+
+        // if this a module or plugin, it may have an associated zip file with it's contents, add that to the test cluster
```
Shouldn't the `PluginBuildPlugin` be doing this?

jakelandis replied:
I'm not sure I follow ?
This section of code is the magic that adds only the required modules/plugins to the test clusters for the OSS YAML tests. This is mostly a copy for how it currently works.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlTestPlugin.java:
```diff
@@ -0,0 +1,126 @@
+        GradleUtils.setupIdeForTestSourceSet(project, yamlTestSourceSet);
+
+        // run test tasks first if they exist since they are presumably faster and less resources intensive
+        Task testTask = project.getTasks().findByName("test");
```
Let's remove this. I don't think it should go here, plus using `findByName` is going to eagerly realize the task.

jakelandis replied:
updated it to be inline with https://github.com/elastic/elasticsearch/pull/57016

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlTestPlugin.java:
```diff
@@ -0,0 +1,126 @@
+        }
+
+        // validation of the rest specification is wired to precommit, so ensure that runs first
+        yamlTestTask.mustRunAfter(project.getTasks().getByName("precommit"));
```
What's creating the `precommit` task in this scenario? Is it some other plugin we're applying?

rjernst replied:
Note https://github.com/elastic/elasticsearch/pull/56926 will make this easier, you shouldn't have to worry about it. Currently that we already have:

```
project.getTasks().named(JavaPlugin.TEST_TASK_NAME).configure(t -> t.mustRunAfter(precommit));
```

jakelandis replied:
I removed this line since #56926 addressed all Test tasks

jakelandis commented:
> Why? The main motivator here is to make these things very explicit. The directory this stuff goes in dictates the kind of test. Supporting such a "hybrid" approach conflicts with that goal. In what scenario would a developer require these test live in the unit testing source set instead of the YAML test one?

Two reasons: 
* To prevent breaking plugin developers that use the YAML test framework (specifically the copy REST artifacts parts need to continue to support the original locations)
* To support our code base until all of the changes are implemented. 

For both cases I can follow up with removal for 8.0 only.

rjernst commented:
>To prevent breaking plugin developers that use the YAML test framework

I don't think we need to wait until 8.0. We have been very explicit to plugin developers that there can be breaks in minor versions.

jakelandis committed:
Merge remote-tracking branch 'upstream/master' into yamlTest_part1

jakelandis committed:
review changes and move core tests

jakelandis committed:
checkstlyle and spotless

jakelandis committed:
minor text changes

jakelandis committed:
remove temp code infavor explicit config

jakelandis committed:
remove rest-resources from build plugin

jakelandis committed:
update the example projects

jakelandis committed:
more missed locations

jakelandis committed:
fix docker tests

jakelandis committed:
fix s3 tests

jakelandis committed:
fix some x-pack tests

jakelandis commented:
@mark-vieira @rjernst - this should be ready for another review. 

Most of the files changed here are due the removal of the rest-resources plugin from the build/standalone plugins in favor of explicitly applying the plugin. As well as the new default target "yamlRestTest" sourceset output for the rest-resources. Moving to be more explicit and updating the default (to the future default) sets up the future changes while preventing breaking tests currently.  

I have also changed the task/source set name to "yamlRestTest" so we can introduce a "javaRestTest" at some point in the future. 

I also moved the core rest tests to a place they make more sense to run as well as to avoid a minor change to the distribution plugin.

jakelandis commented on distribution/archives/build.gradle:
```diff
@@ -330,12 +330,6 @@ configure(subprojects.findAll { it.name == 'integ-test-zip' }) {
-    }
-  }
-
   integTest {
```
I just realized that moving these tests out of this project and into :rest-api-spec has the side effect of changing 
from: running these tests if anything changes (due to the dependsOn assemble) 
to: only running the tests if the any of the actual tests change.   

Do we want to keep the behavior of always running the core rest tests ?

jakelandis committed:
Merge branch 'master' into yamlTest_part1

jakelandis commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into yamlTest_part1

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/RestResourcesExtension.java:
```diff
@@ -32,6 +32,7 @@
 
     final RestResourcesSpec restApi;
     final RestResourcesSpec restTests;
+    private String sourceSetName = "yamlRestTest";
```
Do we have to define this here _and_ in each of the task implementations? Can we have this default just defined on place?

jakelandis replied:
I ended up removing this entirely from the rest resource extension. The which source set to use is now decided by if the YamlRestTestPlugin has been applied. [`6289b00` (#56841)](https://github.com/elastic/elasticsearch/pull/56841/commits/6289b00b7fb71856968e2a6fa543202f4aa8cf7d)

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestTestPlugin.java:
```diff
@@ -0,0 +1,122 @@
+        yamlRestTestTask.setDescription("Runs the YAML based REST tests against an external cluster");
+
+        // setup task dependency
+        if (BuildParams.isInternal()) {
```
The `compile` configuration is deprecated, please use "implementation" instead. Also, no need to construct the configuration name, you can request it via `SourceSet.getImplementationConfigurationName()`.

jakelandis replied:
done [`6289b00` (#56841)](https://github.com/elastic/elasticsearch/pull/56841/commits/6289b00b7fb71856968e2a6fa543202f4aa8cf7d) and [`02afa30` (#56841)](https://github.com/elastic/elasticsearch/pull/56841/commits/02afa305711b7118a64ced93df1607097a5defcb)

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestTestPlugin.java:
```diff
@@ -0,0 +1,122 @@
+        }
+
+        // setup the runner
+        RestTestRunnerTask runner = (RestTestRunnerTask) project.getTasks().getByName(yamlRestTestTask.getName() + "Runner");
```
We should use `RestIntegTestTask.runner()` method to configure the test runner task rather than fetch it by name.

jakelandis replied:
done [`6289b00` (#56841)](https://github.com/elastic/elasticsearch/pull/56841/commits/6289b00b7fb71856968e2a6fa543202f4aa8cf7d) (had to add a getter with `@Internal`)

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestTestPlugin.java:
```diff
@@ -0,0 +1,122 @@
+        // if this a module or plugin, it may have an associated zip file with it's contents, add that to the test cluster
+        boolean isModule = project.getPath().startsWith(":modules:");
+        Zip bundle = (Zip) project.getTasks().findByName("bundlePlugin");
+        if (bundle != null) {
```
We should key this off the application of `PluginBuildPlugin`, not by a particular task. For [example](https://github.com/elastic/elasticsearch/blob/9526c7a4b340782892a2e4a57238f990c2ca2f42/buildSrc/src/main/java/org/elasticsearch/gradle/EnforceDeprecationFailuresPlugin.java#L47).

jakelandis replied:
Unfortunately, `PluginBuildPlugin` is in groovy, and this is in Java which can not reference the groovy side of things.

mark-vieira replied:
No worries, we can do the same by referencing it by id.

```
project.getPluginManager().withPlugin("elasticsearch.esplugin", plugin -> {
    // do stuff
});
```

jakelandis replied:
done [`02afa30` (#56841)](https://github.com/elastic/elasticsearch/pull/56841/commits/02afa305711b7118a64ced93df1607097a5defcb) , thanks even think about using the ID.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestTestPlugin.java:
```diff
@@ -0,0 +1,122 @@
+        GradleUtils.setupIdeForTestSourceSet(project, yamlTestSourceSet);
+
+        // make the new test run after unit tests
+        yamlRestTestTask.mustRunAfter(project.getTasks().named("test"));
```
Can we move this up where we configure other aspects of this task?

jakelandis replied:
done [`6289b00` (#56841)](https://github.com/elastic/elasticsearch/pull/56841/commits/6289b00b7fb71856968e2a6fa543202f4aa8cf7d)

mark-vieira commented on client/rest-high-level/build.gradle:
```diff
@@ -30,6 +30,7 @@ group = 'org.elasticsearch.client'
 archivesBaseName = 'elasticsearch-rest-high-level-client'
 
 restResources {
+  sourceSetName 'test'
```
perhaps to make the migration less cumbersome we should default the rest resources plugin to continue to use the `test` source set, and only set it to `yamlRestTest` when the yaml rest test plugin is applied to a project.

jakelandis replied:
indeed.  I like this better. [`6289b00` (#56841)](https://github.com/elastic/elasticsearch/pull/56841/commits/6289b00b7fb71856968e2a6fa543202f4aa8cf7d)

I had to do a little bit of gymnastics to ensure that configuration that uses source set name is indeed lazily evaluated : [`6289b00` (#56841)](https://github.com/elastic/elasticsearch/pull/56841/commits/6289b00b7fb71856968e2a6fa543202f4aa8cf7d#diff-1e33902b5fc874b367bd346bd6e23790R68) , the dependency from within the `.create` of the task was causing the `.register` of the task to also eagerly evaluate which caused an incorrect source set name since we now set the source set name when the YamlRestTestPlugin is applied.

mark-vieira replied:
I don't see why these gymnastics are necessary. We have the resources plugin define a default value and we override it in the `YamlRestTestPlugin`.

All the stuff in ` project.getPlugins().withType(YamlRestTestPlugin.class)` should just go in `YamlRestTestPlugin` itself.

jakelandis replied:
thanks @mark-vieira I removed the gymnastics as described. However, i did need to leave in the last afterEvalute else we have an ordering requirement between the application of the Java plugin and the RestResources plugin (due to the need to optionally create a dependency from the `processTestResources`)

mark-vieira reviewed (approved):
Left a few minor comments but otherwise LGTM. Thanks for the patience on the review, Jake.

jakelandis committed:
review changes

jakelandis committed:
minor review comments

jakelandis commented:
@mark-vieira thanks for the review.  Would you mind taking a last look at [`6289b00` (#56841)](https://github.com/elastic/elasticsearch/pull/56841/commits/6289b00b7fb71856968e2a6fa543202f4aa8cf7d) , specifically how the RestResourcesPlugin is conditionally applying the source set based on the application of the YamlRestTestPlugin.

jakelandis commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into yamlTest_part1

jakelandis committed:
ensure core specs are copied for x-pack

jakelandis committed:
remove gymnastics

jakelandis committed:
fix checkstyle

jakelandis committed:
fix newly introduced project

jakelandis commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into yamlTest_part1

jakelandis committed:
Merge remote-tracking branch 'upstream/master' into yamlTest_part1

jakelandis commented:
This commit creates a new Gradle plugin to provide a separate task name
and source set for running YAML based REST tests. The only project
converted to use the new plugin in this PR is distribution/archives/integ-test-zip.
For which the testing has been moved to :rest-api-spec since it makes the most
sense and it avoids a small but awkward change to the distribution plugin.

The remaining cases in modules, plugins, and x-pack will be handled in followups.

This plugin is distinctly different from the plugin introduced in #55896 since
the YAML REST tests are intended to be black box tests over HTTP. As such they
should not (by default) have access to the classpath for that which they are testing.

The YAML based REST tests will be moved to separate source sets (yamlRestTest).
The which source is the target for the test resources is dependent on if this
new plugin is applied. If it is not applied, it will default to the test source
set.

Further, this introduces a breaking change for plugin developers that
use the YAML testing framework. They will now need to either use the new source set
and matching task, or configure the rest resources to use the old "test" source set that
matches the old integTest task. (The former should be preferred).

As part of this change (which is also breaking for plugin developers) the
rest resources plugin has been removed from the build plugin and now requires
either explicit application or application via the new YAML REST test plugin.

Plugin developers should be able to fix the breaking changes to the YAML tests
by adding apply plugin: 'elasticsearch.yaml-rest-test' and moving the YAML tests
under a yamlRestTest folder (instead of test)
